source/extensions/common/dynamic_forward_proxy/dns_cache_impl.h

```diff
@@ -117,7 +139,9 @@ class DnsCacheImpl : public DnsCache, Logger::Loggable<Logger::Id::forward_proxy
     Network::ActiveDnsQuery* active_query_{};
   };
 
-  using PrimaryHostInfoPtr = std::unique_ptr<PrimaryHostInfo>;
+  // Hold PrimaryHostInfo by shared_ptr to avoid having to hold the map mutex while updating
+  // individual entries.
```

Comment:
It's not clear to me why this change is required. Expand? If it is required switch to `PrimaryHostInfoSharedPtr`

Reply:
I changed this so we could release the mutex after a map lookup and guarantee that the entry won't be deleted while we're updating it.

Now that I'm thinking more about it, it seems like it might be too strict since the current implementation only adds/removes in the main thread. I think it's probably fine to restore this to unique_ptr, document the constraint that the add/updates are in the main thread, and throw in a few `ASSERT(main_thread_dispatcher_.isThreadSafe())` for good measure. WDYT?

Reply:
Yup SGTM!

Reply:
Restored unique_ptr and added comments/assertions in 3e16973